tommath.h

```diff
@@ -665,7 +662,12 @@ MP_DEPRECATED(mp_prime_rand) mp_err mp_prime_random_ex(mp_int *a, int t, int siz
 mp_err mp_prime_rand(mp_int *a, int t, int size, int flags) MP_WUR;
 
 /* Integer logarithm to integer base */
-mp_err mp_ilogb(const mp_int *a, mp_digit base, mp_int *c) MP_WUR;
+mp_err mp_ilogb(const mp_int *a, uint32_t base, mp_int *c) MP_WUR;
+
+/* c = a**b */
+mp_err mp_expt(const mp_int *a, uint32_t b, mp_int *c) MP_WUR;
```

Comment:
TBH API consistency-wise I would prefer 

```
mp_err mp_expt_u32(const mp_int *a, uint32_t b, mp_int *c) MP_WUR;
mp_err mp_expt(const mp_int *a, const mp_int *b, mp_int *c) MP_WUR;
```

where `mp_expt()` returns `MP_VAL` for b > u32_max and then calls `mp_expt_u32()`

Reply:
this would also close #240

Reply:
#240 also proposes to handle 0, +-1 base special cases for arbitrary large exponent. But I am not convinced that I want that.

Reply:
If we don't add a all bigint mp_expt I am against the suffix since there are no variants of the function.

Reply:
I don't see a reason not to do it if it helps the user write cleaner code... that's the purpose of a library... isn't it?

> #240 also proposes to handle 0, +-1 base special cases for arbitrary large exponent.

that's a good question if we should start adding special cases... if they're not a lot of effort to implement and maintain, why not? that's just another feature and we provide a lot of features which were _a lot_ of effort

> But I am not convinced that I want that.

Why?

> If we don't add a all bigint mp_expt I am against the suffix since there are no variants of the function.

that's why I want the all-bigint version now :-) (so a user can provide its own arbitrary-size `mp_expt()` implementation e.g. the one from #201 if he really sees the need for it)
also it's consistent and I like consistency

Reply:
I don't think mp_expt is a good function since it will fail on most inputs and it will only work on very specific inputs.

* will work for <= UINT32_MAX (or even << UINT32_MAX for large bases)
* won't work for > UINT32_MAX, except for the base is 0,+-1

This kind of special handling is better done by the user. If the user wants it, it is very easy to implement by extracting the low bits as we discussed in #240.

> that's why I want the all-bigint version now :-)

If you want it feel free to add it - I won't complain too much ;) But I think the API proposed in this PR is the most reasonable one. I guess there are more functions which you could generalise (in a bad way) by promoting arguments from int to mp_int. Do you want these too?

Reply:
To say it in different words - the type of the function `mp_expt(const mp_int*, const mp_int*, mp_int*)` makes a promise that it works on more inputs than it does really or in practice. Therefore I don't like it.

Reply:
> I guess there are more functions which you could generalise (in a bad way) by promoting arguments from int to mp_int. Do you want these too?

good point! now I'm not so sure anymore 

hmm ...I just scrolled through the API and there are not many math operations which could be promoted ... and they're basically all the exact same case as for this discussion... (I probably missed some operations!?)

```
/* c = a / 2**b, implemented as c = a >> b */
mp_err mp_div_2d(const mp_int *a, int b, mp_int *c, mp_int *d) MP_WUR;

/* c = a * 2**b, implemented as c = a << b */
mp_err mp_mul_2d(const mp_int *a, int b, mp_int *c) MP_WUR;

/* c = a mod 2**b */
mp_err mp_mod_2d(const mp_int *a, int b, mp_int *c) MP_WUR;

/* computes a = 2**b */
mp_err mp_2expt(mp_int *a, int b) MP_WUR;
```

Reply:
Yes and we should neither promote those nor should we promote `mp_expt`.